sc2/bot_ai.py

```diff
@@ -207,35 +208,26 @@ def expansion_locations(self) -> Dict[Point2, Units]:
         resources (mineral field and vespene geyser) as value.
         """
 
-        # Idea: create a group for every resource, then merge these groups if
-        # any resource in a group is closer than a threshold to any resource of another group
+        # - Create a graph of resource nodes, connect nearby nodes on the graph
 
         # Distance we group resources by
         resource_spread_threshold = 8.5
         geysers = self.vespene_geyser
-        # Create a group for every resource
-        resource_groups = [
-            [resource]
-            for resource in self.resources
-            if resource.name != "MineralField450"  # dont use low mineral count patches
-        ]
-        # Loop the merging process as long as we change something
-        merged_group = True
-        while merged_group:
-            merged_group = False
-            # Check every combination of two groups
-            for group_a, group_b in itertools.combinations(resource_groups, 2):
-                # Check if any pair of resource of these groups is closer than threshold together
-                if any(
-                    resource_a.distance_to(resource_b) <= resource_spread_threshold
-                    for resource_a, resource_b in itertools.product(group_a, group_b)
-                ):
-                    # Remove the single groups and add the merged group
-                    resource_groups.remove(group_a)
-                    resource_groups.remove(group_b)
-                    resource_groups.append(group_a + group_b)
-                    merged_group = True
-                    break
+        resource_graph = networkx.Graph()
+        resources = [ resource for resource in self.resources if resource.name != 'MineralField450']
```

Comment:
Make this a generator: `resources = (resource for resource in self.resources if resource.name != 'MineralField450')`